core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -300,6 +331,20 @@ public synchronized void close() throws IOException {
     }
   }
 
+  /**
+   * Closes the context, update configuration from meta master, then re-initializes the context.
+   *
+   * @throws IOException when failed to close the context or update configuration
+   */
+  public synchronized void reinit() throws IOException {
```

Comment:
how does this affect outstanding RPCs?

Reply:
When closing, all grpc channels and underlying connections will be closed, and each grpc channel has its ID, a response from the rpc sent by old client will not be received by the new client.

Reply:
Will config propagation lead to user-facing errors, or will we perform retries, similar to what we do when the primary master fails over to a standby

Reply:
It will lead to user-facing errors

Reply:
I don't think this acceptable, what would it take to prevent updating conf from auto-failing outstanding jobs?

Reply:
Retrying will cause inconsistencies, e.g. when writing a file, if previously it's CACHE but the conf is updated to THROUGH, then part of the file could be only cached while the remaining is persisted.

Reply:
@calvinjia is it acceptable to fail all outstanding rpcs when conf is updated? I could see this being ok for compute frameworks if they clean up and retry on failure, but not all clients will do this. We put a lot of effort into making master failover transparent to compute, and this seems similar.

@cheng-chang retrying doesn't necessarily cause inconsistencies - we could allow the file write to complete using the configuration it was started with.

Reply:
For the stateful interactions (file in/out streams), could we keep the configuration constant (ie. once a stream is created, it always has that configuration). For everything else I think we can retry.